RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,823 @@
+
+This is also a very complicated feature, with lots of edge cases, that has significantly reduced utility if any of the following are true:
+
+* There is insufficient flow analysis such that F# programmers are asserting non-nullability all over the place just to avoid unnecessary warnings
+* Existing, working code is infected without the programmer explicitly opting into this new behavior
+* The rest of the .NET ecosystem simply does not adopt non-nullability, thus making adornments the new normal
+
+Although we cannot control the third point, ensuring that the implementation is complete and up to spec with respect to what C# also does (while accounting for F#-isms) is the only way to do our part in helping the ecosystem produce less nulls.
+
+Additionally, we now have the following possible things to account for with respect to accessing underlying data:
+
+* Nullable reference types
+* Option
+* ValueOption
+* Choice
+* Result
+
+These are a lot of ways to send data to various parts of the system, and there is a danger that people could use nullable reference types for more than just the boundary levels of their system.
+
+### Unsoundness
+
+This feature is inherently unsound. For starters, warnings don't prevent people from flowing `null` through their code like errors do. Compare this with the F# option type, which offers zero way to attempt to coerce `None` into a `Some x`.
+
+More subtly, there is no way to extract nullability information with reflection today without a change in the .NET runtime. This means that it can be impossible to know, at least by reflection, if the type of something is actually nullable or non-nullable. This means that reflection-based libraries (such as JSON.NET) may be able to define a contract that they cannot fulfill; i.e., make an API's return type a `'T` when it could still be `null`.
+
+Finally, the "handles null" attribute can be trivially abused, and it may be likely that the F# compiler could be "tricked" into thinking something will no longer be `null`.
+
+All of this means that this feature is entrusting acting in good faith on the greater .NET ecosystem, rather than preventing acting in bad faith with compile-time guarantees.
+
+## Alternatives
+[alternatives]: #alternatives
+
+The primary alternative is to simply not do this.
+
+That said, it will become quite evident that this feature is necessary if F# is to continue advancing on the .NET platform. Despite originating as a C# feature, this is fundamentally a **platform-level shift** for .NET, and in a few years, will result in .NET looking very different than it is today.
+
+### Syntax
+
+#### Question mark nullability annotation
+
+```fsharp
+let ns: string? = "hello"
+
+type C<'T?>() = class end
+```
+
+Issue: looks horrible with F# optional parameters
```

Comment:
I find question mark the best option because it's the same as in C# and Kotlin and because it's far shorted than `ty | null`. Yes, it looks horrible with F# optional parameters, but they are used quite rarely and this should not get in the way of accepting `?` syntax.

Reply:
I've no great desire to make either `string | null` or `string?` used heavily in F# code.